data/query.js

```diff
@@ -0,0 +1,66 @@
+/**
+ * External Dependencies
+ */
+import { isFunction } from 'lodash';
+
+/**
+ * WordPress Dependencies
+ */
+import { Component } from '@wordpress/element';
+
+const createQueryHigherOrderComponent = ( client ) => ( mapPropsToQuery, mapPropsToVariables = () => ( {} ) ) => ( WrappedComponent ) => {
+	return class GraphQueryComponent extends Component {
+		constructor() {
+			super( ...arguments );
+			this.state = {
+				data: null, errors: null,
+			};
+		}
+
+		componentDidMount() {
+			this.buildQuery( this.props );
+			this.request();
+		}
+
+		componentWillUnmount() {
+			this.cancelRequest();
+		}
+
+		componentWillReceiveProps( newProps ) {
+			this.buildQuery( newProps );
+			this.request();
+		}
+
+		buildQuery( props ) {
+			if ( isFunction( mapPropsToQuery ) ) {
+				this.query = mapPropsToQuery( props );
+			} else {
+				this.query = mapPropsToQuery;
+			}
+			this.variables = mapPropsToVariables( props );
+		}
+
+		cancelRequest() {
+			if ( this.unsubscribe ) {
+				this.unsubscribe();
+			}
+		}
+
+		request() {
+			this.cancelRequest();
+			const query = client.query( { query: this.query, variables: this.variables } );
+			const observer = query.subscribe( ( results ) => {
+				this.setState( results );
```

Comment:
In your example in the pull request description the data is accessed using `props.data`, but in the higher order component you only set the state. How does this work?

Reply:
Yes, actually a GraphQL result is always like this `{ data, errors }` so we'll end up with these two props. 

For convenience we could add a callback to "polish" the props like suggested by @aduth. `mapResultsToProps`